rclcpp/include/rclcpp/executor.hpp

```diff
@@ -570,6 +572,55 @@ class Executor
   virtual void
   spin_once_impl(std::chrono::nanoseconds timeout);
 
+protected:
+  // Implementation details, used by spin_until_complete and spin_for.
+  // Previous implementation of spin_until_future_complete.
+  template<typename ConditionT, typename DurationT>
+  FutureReturnCode
+  spin_until_complete_impl(ConditionT condition, DurationT timeout)
+  {
+    auto end_time = std::chrono::steady_clock::now();
+    std::chrono::nanoseconds timeout_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      timeout);
+    if (timeout_ns > std::chrono::nanoseconds::zero()) {
+      end_time += timeout_ns;
+    }
+    std::chrono::nanoseconds timeout_left = timeout_ns;
+
+    // Preliminary check, finish if conditon is done already.
+    if (condition()) {
+      return FutureReturnCode::SUCCESS;
+    }
+
```

Comment:
This implementation is racy. If the condition switches after this line of code, and an infinite timeout is given, we got a dead lock.

Reply:
Currently the function is named spin_until_complete. Therefore I would expect, I can pass it a std::future or any condition, and it will do the correct thing, and protect me from this kind of races. 

This is clearly not the case. I think it is not even possible with the current API. 
To solve this, we would also need to pass a std::unique_lock, like done here:
https://en.cppreference.com/w/cpp/thread/condition_variable/wait

Reply:
Wait, I don't see how this can cause an _unwanted_ infinite wait/spin.

It is assumed that `condition()` is initially `false`. If it is `true` from the get-go, then it just returns without spinning. Otherwise, it spins once before checking if `condition()` is `true`, otherwise it spins once again, and returns if it times out. If `condition()` never changes to `true` and `timeout` is negative, then yes it will spin infinitely, but that's intentional.

Am I missing something?

Reply:
That still feels like user error to me. @wjwwood thoughts?

Reply:
I am more on the side of bad documentation of the function, and bad API, as you can pass in anything here.

I am pushing this point, as this is a public function, that is even promoted in the tutorials (the future version). But using it is not trivial if you want to get deterministic results. Whatever you pass in needs to explicitly wake up the wait after it changes the result of the condition to true. And doing this requires knowledge of rather internal stuff.

Reply:
FYI we will not be changing downstream calls to use this new API (https://github.com/ros2/rclcpp/pull/2475#discussion_r1550080593), so I've updated https://github.com/ros2/ros2_documentation/pull/4252 to only mention this change and not change any tutorials.

Reply:
I think it's a documentation issue, it's expected that if you ask for it to wait until a condition is complete, then you need your condition to change before waking the wait set. This has all the same considerations as passing a predicate to a condition variable, except that a condition variable can spurious wake up, so even if you do something wrong (e.g. have the predicate change after interrupting, resulting in a data race), it might not result in a deadlock.

As to it being bad API, I don't see how we could enforce that people write their condition function such that it works correctly.

The only thing we could do (that I can think of) is to restrict the API to only take futures that are produced and managed by rclcpp itself (that way we can ensure they're coded correctly), but that's even more restrictive than what is there now.

Or were you thinking of something else?

Is it sufficient to document that a condition function (a.k.a. a predicate) must change its answer before waking the wait set?

Reply:
I see tow ways:

change the function to
```c++
  spin_until_complete(
    const std::function<bool(void)> & condition,
    std::chrono::duration<TimeRepT, TimeT> timeout = std::chrono::duration<TimeRepT, TimeT>(-1),
     std::chrono::duration<TimeRepT, TimeT> condition_check_interval = std::chrono::millisecons(10))
``` 
And have the condition checked in this interval.

You could create an object like this
```c++
class WaitCondition
{
    WaitCondition();

    void stop_spin();
};
```

And change the function to
```c++
  spin_until_complete(
    const std::function<bool(WaitCondition &cond)> & condition,
    std::chrono::duration<TimeRepT, TimeT> timeout = std::chrono::duration<TimeRepT, TimeT>(-1))
```  

Now, that I tried to use this interface I am confused of how it is intended to work. @christophebedard  can you provide an example of a function, you would pass into this interface ?

Reply:
Regarding my question
Is the condition actively changed by an external event / callback ? 
It is supposed to be changed by a second thread in some way ?

Reply:
>  Is the condition actively changed by an external event / callback ?

Yes, that's the implication I think. You could either have a condition that's affected by some callback in the executor or you could have a condition that is completely decoupled with the tasks in the executor, but that could lead to a dead lock as you suggested. If you wanted to do the latter it should be probably be done using a guard condition that you add to the executor to ensure the condition's state is checked after it has been changed by triggering the guard condition after changing the condition.

>  It is supposed to be changed by a second thread in some way ?

As I said, yes, you could but you'd need to ensure the wait set is notified after the condition had changed, which you could do from a second thread using a guard condition.

> can you provide an example of a function, you would pass into this interface ?

I think the simplest example is the implementation of the syntactic sugar function `spin_until_future_complete()`, something like:

```c++
  template<typename FutureT, typename TimeRepT = int64_t, typename TimeT = std::milli>
  FutureReturnCode
  spin_until_future_complete(
    const FutureT & future,
    std::chrono::duration<TimeRepT, TimeT> timeout = std::chrono::duration<TimeRepT, TimeT>(-1))
  {
    const auto condition = [&future]() {
        return future.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
      };
    return spin_until_complete(condition, timeout);
  }
```

But even in this case (and we should document that), the future given to this function should have it's promise set before notifying the wait set (and it must notify the wait set). The obvious use case for this is using a future created by rclcpp itself, e.g. `client->send_request_async(...)` which returns a future which is completed in an executor callback. But there are others, like completing a future from a callback in a test to spin until a callback was called by the executor.

Reply:
That would be really helpful, I think. I can't think of way to misuse that API right now, definitely safer.

I'm wondering if that would an acceptable API for the people who pushed for this change in the first place? @SteveMacenski I think you were one of the folks asking for the change (but I'd be happy to hear from others too), what do you think about an API like the above where you get an object that can stop the executor rather than passing a callback function?

The future API was really similar (except a little less safe since the user needed to make sure not to fulfill the promise before changing their data), so I'm afraid it will suffer the same inconveniences.

Personally I like it better.

Also, normally I'd say "yes and ..." ask for it to be a follow up pr, but I think we should probably not have this new `spin_until_complete()` API if we choose to use this newer proposal.

Reply:
> what do you think about an API like the above where you get an object that can stop the executor rather than passing a callback function?

I have 2 general goals:
- Make the APIs easier to use / understand without exposing the technical details under the hood for the Jr Eng / researcher / student that shouldn't be required to learn the details, and 
- Lessen the strict requirement that these are specifically futures -- which also has the benefit of making programs lines shorter since `spin_until_future_complete()` takes up alot of space in line writing application code with this in there day-in day-out. It seems like a minor thing, but with the amount of actions I touch, it would actually be a QoL increase. But I also had some ideas for other non-future things I'd like to spin, but the details of that old dream are escaping me for the moment. 

So, I wouldn't support this change if it requires users in their application code to "own" the construction and management of some new Condition object in order to pass to the spin function from the output of a service / action client -- that's making my first goal even more cumbersome with a further level of indirection to understand. 

If however that's handled under the hood for users as part of the rclcpp API so its hidden from them, no issue. We can have multiple implementations of `spin_until_complete` which takes in different types of inputs, that's largely the purpose of this API change -- decouple from futures and allow for other options.

Reply:
Ok, with that in mind, why don't we move forward with this change and consider adding a version like you described @jmachowinski soon (after the API freeze though) for the next release. We can also consider if making this "wait condition" style api the recommended way and even deprecating the callback later.

We just don't have time to hash out a new approach right now, and while still error prone, as you pointed out, the callback function isn't really any more hazardous than what we have now, and it satisfies @SteveMacenski's concerns.

Reply:
@wjwwood to quote your here :
"The thing is that if we merge this just because Monday is coming up, then we have to come back and undo what we merged through a deprecation cycle, which is very undesirable."

I still haven't seen any example of how this API would be used in a real word application.
@SteveMacenski you want this API for "Make the APIs easier to use / understand without exposing the technical details under the hood for the Jr Eng / researcher / student that shouldn't be required to learn the details, and"

The currently proposed API does not reach this goal, as you need to be aware that you need to wake the wait, and to do this, you need to either have a known side effect of another API (which can change by any internal implementation change) or explicit manage a guard condition.

Reply:
That is true, and I stand by what I said, we don't want to merge things that we think will be obsolete immediately.

However, I believe that the situation where you're pulling my quote from (https://github.com/ros2/rclcpp/pull/2466#discussion_r1564553671) and this one are quite different. First, it's not clear to me that what you've proposed here would actually replace the desire for the new API in this pr, so we may or may not deprecate what this pr adds even with the approach you suggested implemented and merged later. Whereas in the other situation, it's more of an A or B style design question. If we choose A now, we'll have a mess switching direction to B in the future, and it would never make sense to have both. Therefore it's more important to act when we're sure it's the right approach, unlike here, where adding a new function is by comparison low risk.

Reply:
I still believe, that the currently proposed API is a to big pitfall for end users.
E.g.
```c++
spin_until_complete([end = std::chrono::system_clock::now() + std::duration::Seconds(1)] ()
{
return end < std::chrono::system_clock::now();
});
```  
will not work as expected. I also don't really see how it will make the live of a user easier in its current state.

I also proposed 
```c++
spin_until_complete(
    const std::function<bool(void)> & condition,
    std::chrono::duration<TimeRepT, TimeT> timeout = std::chrono::duration<TimeRepT, TimeT>(-1),
     std::chrono::duration<TimeRepT, TimeT> condition_check_interval = std::chrono::millisecons(10))
``` 
This approach, which is trivial to implements, and would make the situation way better.

Reply:
> I also don't really see how the API will make the live of a user easier in its current state.

I also don't link the callback approach is really that helpful (but I think it can result in less code), for that you should talk with @SteveMacenski. But I still think it's no better or worse than using futures.

> this approach, which is trivial to implements, and would make the situation way better.

Having the check interval (essentially polling) is bandaid, really the condition should always be attached to some activity that causes the spin to interrupt in an event driven way. If you want something like the check interval, you can just use the timeout argument and loop the function, or you could use an empty timer, etc. etc.

Most of the time this is used to wait for a Service Client to receive a response to its request, or something similar, and for that this function will work fine, and in fact the future-based version of this worked fine.

---

@christophebedard @fujitatomoya can you guys comment on whether or not to delay this until after the release or not?

Reply:
>> I also don't really see how the API will make the live of a user easier in its current state.
>
> I also don't think the callback approach is really that helpful (but I think it can result in less code), for that you should talk with @SteveMacenski. But I still think it's no better or worse than using futures.

Just to be clear, I picked up this set of PRs mainly to finally push them across the finish line. I personally do not have a need for this API in itself, so I'd also like to hear what Steve (or any other potential user) has to say here.

> @christophebedard @fujitatomoya can you guys comment on whether or not to delay this until after the release or not?

At this point I'd prefer to just push it to make sure to get it right.